### ISSUE TITLE
Using flutter_blue_plus

### ISSUE THREAD

**Issue by ivanhercaz** [ogabrielinacio/esp_provisioning_ble#9]
Hi, @ogabrielinacio!
I am working on the implementation of the  `ProvTransport` using `flutter_blue_plus` and Riverpod, although it is a WIP and, although I have some code, it isn't working fine yet. When I have it done I will release my progress.

In addition, I am fighting with an error because at this moment I am not being able to establish the session with the device. I am trying to debug the error. The result of `establishSession()` is `EstablishSessionStatus.disconnected`, so I am using a `whenComplete()` to get more information (`catchError` doesn't catch the error):

```
DisconnectReason{platform: ErrorPlatform.android, code: 8, CONNECTION_TIMEOUT}
```

I read in the README that you have a TODO about testing the package and the example with `flutter_blue_plus`. Have you got any progress on it? And, of course, some progress you can share ❤️.

Thanks in advance! And excuse for not provide more information nor code. As soon as possible I will do it!

**Comment by ogabrielinacio** [comment]
Hi @ivanhercaz , long time no see! I hope you're doing well.

I'm really glad you're testing the package with flutter_blue_plus. This is something I really want to do in the future, but unfortunately I haven't had the time yet, so I don't have any progress to share.

If possible, I would be very happy if you could share this code with the community and open a PR here in this package. You are one of the most deserving people, and if you want we can create a dev branch and work on top of it with your implementation.

If the code cannot be shared due to business reasons, I understand completely. In that case, we can still do the same with just the package implementation on top of it, and using the example provided in the package.

**Comment by ivanhercaz** [comment]
> Hi @ivanhercaz , long time no see! I hope you're doing well.
Yes! There were complicated situations, but now I am trying to recover the lost work time. I hope you're doing well too!

> I'm really glad you're testing the package with flutter_blue_plus. This is something I really want to do in the future, but unfortunately I haven't had the time yet, so I don't have any progress to share.

Oh, don't worry! I am testing not just with `flutter_blue_plus` but also with Riverpod, so two in one, although my last experiment was with Riverpod too, this one is using Riverpod generators because they are really useful.

> If possible, I would be very happy if you could share this code with the community and open a PR here in this package. You are one of the most deserving people, and if you want we can create a dev branch and work on top of it with your implementation.

Of course! When I have something minimal working fine I will share here, but now I am dealing with a `EstablishSessionStatus.keymismatch` and yes, I confirm multiple times the POP is the correct one, confirmed even in the firmware. The steps I am doing are basically:

```dart
var transport = TransportBLE(deviceSelected);
await transport.connect();

var prov = EspProv(
  transport: transport,
  security: Security1(
      pop: 'Secret POP!'),
);
await prov.establishSession();
```

But as I said, I get that `keymismatch` that is making me crazy haha! I will continue researching about it.
 
> If the code cannot be shared due to business reasons, I understand completely. In that case, we can still do the same with just the package implementation on top of it, and using the example provided in the package.

Of course! When I have bit of time I would like to provide the next two things to this package:

1. An independent example using Riverpod and `flutter_blue_plus`.
2. An implementation with `flutter_blue_plus` with the current example.

Both can be contributions I could do in the near-future to this repository or, just the one you prefer, and the another I could share in an independent repository.

**Oftopic**: have you though about using Codeberg instead of Github? It doesn't need an answer hehe, I don't want to convert this thread in that discussion, just make you curious about an interesting forge alternative to Github ❤️ .

**Comment by ivanhercaz** [comment]
Update: I think I am implementing erroneously the Future<Uint8List> sendReceive(String? epName, Uint8List? data)`:

https://github.com/ogabrielinacio/esp_provisioning_ble/blob/e0948a571a4b77b8688e9fe3ae5a04aaa8ea4387/example/lib/src/features/services/transport_ble.dart#L56

I will notice you if I achieve to make it works!

**Comment by ivanhercaz** [comment]
> Great!! i will creat a new dev branch to work with flutter_blue_plus and the package, feel free to make PR, 
> 
> branch: 
> 
> https://github.com/ogabrielinacio/esp_provisioning_ble/tree/dev-flutter_blue_plus

Nice! Really nice! 

> I just added the package to the project in the example part. It's quite late here in my country right now, so I'm not going to test it it today, I took a look at the flutter_blue_plus documentation and it seems to be a slightly more complex library, thus giving us more possibilities, their documentation seems to be very good.
> 
> Unfortunately I won't be able to work on the project this weekend.

Please, don't worry about that part! I will make the PR for the dev branch as soon as possible, maybe this weekend or the next week. I already have the implementation and I just need to test it with the example. Then we can make a peer review of the code. I am sure you can save time in that way and invest in other issues/tasks/life and in the meantime, reviewing I will learn a lot because I am sure you are going to make really good suggestions to improve the implementation.

> answering your Oftopic question : Yes, I know Codeberg, but I chose to host the package on github for visibility, since in the context of hosting git projects it is one of the most used

Thank you for your answer! I have another thinking about it. If in the future you want to migrate to Codeberg and need help, just tell me!

**Comment by ivanhercaz** [comment]
Excuse me the delay, @ogabrielinacio! I have had hard days but I think this weekend I could open the merge request with the transport implementation and test it with the example.

**Comment by adamgasiorek** [comment]
Hey @ivanhercaz @ogabrielinacio any update about this ? We are waiting for this 😃 🥳

**Comment by ogabrielinacio** [comment]
Hey @adamgasiorek  , sorry for the delayed response. This semester, I'll focus on fixing the bugs in the package first, and after that, I'll work on the new feature for flutter_blue_plus. I've been really busy and couldn't work on the package before—sorry about that.

**Comment by ivanhercaz** [comment]
> Hey @ivanhercaz @ogabrielinacio any update about this ? We are waiting for this 😃 🥳

Hi @adamgasiorek! I am really sorry because I have not seen that until now. I think in some moment I marked all notifications as readed or something else, and I doesn't see this one.

I started working on it but I have not had time to check, review and push the changes into a PR.

> I'll work on the new feature for flutter_blue_plus

@ogabrielinacio, I already have work done on this, what do you think about the next approach?

1. I review my work done, create an implementation of `ProvTransport` using `flutter_blue_plus`, commit and push into a branch with a specific PR to review the work.
2. You can review it and implement the current BLOC approach you are using for the documentation and the current example. I have not used BLOC, what I usually use it is Riverpod, but I think we shouldn't change this because we have not time to do it in a good way at this moment, so the better approach may be reuse the BLOC structure you have.

What do you think?

And, one more time, excuse me the delay answering your comment, @adamgasiorek!
